libpages/server.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libpages
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/keybase/kbfs/libkbfs"
+	"go.uber.org/zap"
+	"golang.org/x/crypto/acme"
+	"golang.org/x/crypto/acme/autocert"
+)
+
+// ServerConfig holds configuration parameters for Server.
+type ServerConfig struct {
+	AutoDirectHTTP     bool
+	DomainWhitelist    []string
+	UseStaging         bool
+	Logger             *zap.Logger
+	UseDiskCacheForDev bool
+}
+
+const fsCacheSize = 2 << 15
+
+// Server handles incoming HTTP requests by creating a Root for each host and
+// serving content from it.
+type Server struct {
+	config     ServerConfig
+	kbfsConfig libkbfs.Config
+
+	fsCache *lru.Cache
+
+	whiteList     map[string]bool
+	whiteListOnce sync.Once
+}
+
+func (s *Server) getFS(
+	ctx context.Context, root Root) (http.FileSystem, error) {
+	fsCached, ok := s.fsCache.Get(root)
+	if !ok {
+		fs, err := root.MakeFS(ctx, s.config.Logger, s.kbfsConfig)
+		if err != nil {
+			return nil, err
+		}
+		s.fsCache.Add(root, fs)
+		return fs, nil
+	}
+	fs, ok := fsCached.(http.FileSystem)
+	if !ok {
+		fs, err := root.MakeFS(ctx, s.config.Logger, s.kbfsConfig)
+		if err != nil {
+			return nil, err
+		}
+		s.fsCache.Add(root, fs)
+		return fs, nil
+	}
+	return fs, nil
+}
+
+func (s *Server) handleError(w http.ResponseWriter, err error) {
+	// TODO: have a nicer error page for configuration errors?
+	switch err.(type) {
+	case nil:
+	case ErrKeybasePagesRecordNotFound:
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	case ErrKeybasePagesRecordTooMany:
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	case ErrInvalidKeybasePagesRecord:
+		w.WriteHeader(http.StatusPreconditionFailed)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+}
+
+// ServeHTTP implements the http.Handler interface.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	root, err := LoadRootFromDNS(s.config.Logger, r.Host)
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+	fs, err := s.getFS(r.Context(), root)
```

Comment:
We should probably a new debug tag into the ctx for each of `ServeHTTP` call, e.g. with `libkbfs.CtxWithRandomIDReplayable`.